sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySchemaIOProvider.java

```diff
@@ -17,199 +17,63 @@
  */
 package org.apache.beam.sdk.io.gcp.bigquery;
 
-import com.google.api.services.bigquery.model.TableRow;
-import com.google.auto.service.AutoService;
-import java.io.Serializable;
-import java.util.HashMap;
+import java.util.Collections;
+import java.util.List;
 import org.apache.beam.sdk.annotations.Experimental;
 import org.apache.beam.sdk.annotations.Internal;
-import org.apache.beam.sdk.schemas.Schema;
-import org.apache.beam.sdk.schemas.Schema.FieldType;
-import org.apache.beam.sdk.schemas.io.SchemaIO;
-import org.apache.beam.sdk.schemas.io.SchemaIOProvider;
-import org.apache.beam.sdk.schemas.transforms.Convert;
-import org.apache.beam.sdk.transforms.PTransform;
-import org.apache.beam.sdk.values.PBegin;
-import org.apache.beam.sdk.values.PCollection;
-import org.apache.beam.sdk.values.PDone;
-import org.apache.beam.sdk.values.Row;
-import org.checkerframework.checker.nullness.qual.Nullable;
+import org.apache.beam.sdk.io.gcp.bigquery.schematransform.BigQuerySchemaIOConfiguration;
+import org.apache.beam.sdk.io.gcp.bigquery.schematransform.BigQuerySchemaTransform;
+import org.apache.beam.sdk.schemas.transforms.SchemaTransform;
+import org.apache.beam.sdk.schemas.transforms.TypedSchemaTransformProvider;
 
 /**
- * An implementation of {@link SchemaIOProvider} for reading and writing to BigQuery with {@link
- * BigQueryIO}. For a description of configuration options and other defaults, see {@link
- * BigQuerySchemaIOProvider#configurationSchema()}.
+ * An implementation of {@link TypedSchemaTransformProvider} for reading and writing to BigQuery
+ * with {@link BigQueryIO}.
  *
- * <p>This transform is still experimental, and is still subject to breaking changes.
+ * <p><b>Internal only:</b> This is actively being worked on and will likely change. We provide no
+ * backwards compatibility guarantees, and it should not be implemented outside the Beam repository.
  */
 @Internal
 @Experimental
-@AutoService(SchemaIOProvider.class)
```

Comment:
We need one of these for read and write, right? Shouldn't that be in the name?

Reply:
Typically I've seen the SchemaProvider + the SchemaTransform/IO in one file. So we'd have two sets of those. I don't know if that fits with the classes here or not though.